ios/AppLovinMAXAdView.m

```diff
@@ -0,0 +1,189 @@
+// The following properties are updated from RN layer via the view manager
+@property (nonatomic, copy) NSString *adUnitId;
+@property (nonatomic, weak) MAAdFormat *adFormat;
+@property (nonatomic, copy, nullable) NSString *placement;
+@property (nonatomic, copy, nullable) NSString *customData;
+@property (nonatomic, assign, readonly, getter=isAdaptiveBannerEnabled) BOOL adaptiveBannerEnabled;
+@property (nonatomic, assign, readonly, getter=isAutoRefresh) BOOL autoRefresh;
+
+@end
+
+@implementation AppLovinMAXAdView
+
+- (void)setAdUnitId:(NSString *)adUnitId
+{
+    // Ad Unit ID must be set prior to creating MAAdView
+    if ( self.adView )
+    {
+        [[AppLovinMAX shared] log: @"Attempting to set Ad Unit ID %@ after MAAdView is created", adUnitId];
+        return;
+    }
+    
+    _adUnitId = adUnitId;
+    
+    [self attachAdView];
+}  
+
+- (void)setAdFormat:(NSString *)adFormat
+{
+    // Ad format must be set prior to creating MAAdView
+    if ( self.adView )
+    {
+        [[AppLovinMAX shared] log: @"Attempting to set ad format %@ after MAAdView is created", adFormat];
+        return;
+    }
+    
+    if ( [@"banner" isEqualToString: adFormat] )
+    {
+        _adFormat = DEVICE_SPECIFIC_ADVIEW_AD_FORMAT;
+    }
+    else if ( [@"mrec" isEqualToString: adFormat] )
+    {
+        _adFormat = MAAdFormat.mrec;
+    }
+    else
+    {
+        [[AppLovinMAX shared] log: @"Attempting to set an invalid ad format of \"%@\" for %@", adFormat, self.adUnitId];
+        return;
+    }
+    
+    [self attachAdView];
+}  
+
+- (void)setPlacement:(NSString *)placement
+{
+    _placement = placement;
+    
+    if ( self.adView )
+    {
+        self.adView.placement = placement;
+    }
+}
+
+- (void)setCustomData:(NSString *)customData
+{
+    _customData = customData;
+    
+    if ( self.adView )
+    {
+        self.adView.customData = customData;
+    }
+}
+
+- (void)setAdaptiveBannerEnabled:(BOOL)adaptiveBannerEnabled
+{
+    _adaptiveBannerEnabled = adaptiveBannerEnabled;
+    
+    if ( self.adView )
+    {
+        [self.adView setExtraParameterForKey: @"adaptive_banner" value: adaptiveBannerEnabled ? @"true" : @"false"];
+    }
+}
+
+- (void)setAutoRefresh:(BOOL)autoRefresh
+{
+    _autoRefresh = autoRefresh;
+    
+    if ( self.adView )
+    {
+        if ( autoRefresh )
+        {
+            [self.adView startAutoRefresh];
+        }
+        else
+        {
+            [self.adView stopAutoRefresh];
+        }
+    }
+}
+
+- (void)attachAdView
+{
+    if ( ![self.adUnitId al_isValidString] )
+    {
+        [[AppLovinMAX shared] log: @"Attempting to attach MAAdView without Ad Unit ID"];
+        return;
+    }
+    
+    // Re-assign in case of race condition
+    NSString *adUnitId = self.adUnitId;
+    MAAdFormat *adFormat = self.adFormat;
+    
+    // Run after 0.25 sec delay to allow all properties to set
+    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) (NSEC_PER_SEC/4)), dispatch_get_main_queue(), ^{
```

Comment:
```suggestion
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t) (0.25 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
```

Reply:
Why I suggest this is due to readability - traditionally the value is the <# of seconds> * NSEC_PER_SEC.

We're aiming for 0.25s, so we should just specify that. It's more straightforward, as opposed to "let's divide `NSEC_PER_SEC` by 4".

We also can't use this if we want to do for example 0.037 seconds or something that's not easily divisible

Reply:
Got it, updated.